chia/util/db_wrapper.py

```diff
@@ -29,3 +33,96 @@ async def rollback_transaction(self):
+        self._read_connections = asyncio.Queue()
+        self._write_connection = connection
+        self._lock = asyncio.Lock()
+        self.db_version = db_version
+        self._num_read_connections = 0
+        self._in_use = {}
+        self._current_writer = None
+
+    async def close(self) -> None:
+        while self._num_read_connections > 0:
+            await (await self._read_connections.get()).close()
+            self._num_read_connections -= 1
+        await self._write_connection.close()
+
+    @contextlib.asynccontextmanager
+    async def write_db(self) -> AsyncIterator[aiosqlite.Connection]:
+        task = asyncio.current_task()
+        assert task is not None
+        if self._current_writer == task:
+            # we allow nesting writers within the same task
+            yield self._write_connection
+            return
+
+        async with self._lock:
+
+            global savepoint_name
+            name = f"s{savepoint_name}"
+            savepoint_name += 1
+
+            await self._write_connection.execute(f"SAVEPOINT {name}")
+            try:
+                self._current_writer = task
+                yield self._write_connection
+            except BaseException:
+                await self._write_connection.execute(f"ROLLBACK TO {name}")
+                # rollback to a savepoint doesn't cancel the transaction, it
+                # just restarts it at its initial state. We want to cancel it
+                # here, so we release it as well
+                await self._write_connection.execute(f"RELEASE {name}")
+                raise
+            finally:
+                self._current_writer = None
+            await self._write_connection.execute(f"RELEASE {name}")
+
+    @contextlib.asynccontextmanager
+    async def read_db(self) -> AsyncIterator[aiosqlite.Connection]:
+        # there should have been read connections added
+        assert self._num_read_connections > 0
+
+        # we can have multiple concurrent readers, just pick a connection from
+        # the pool of readers. If they're all busy, we'll wait for one to free
+        # up.
+        task = asyncio.current_task()
+        assert task is not None
+
+        # if this task currently holds the write lock, use the same connection,
+        # so it can read back updates it has made to its transaction, even
+        # though it hasn't been comitted yet
+        if self._current_writer == task:
+            # we allow nesting writers within the same task
+            yield self._write_connection
+            return
+
+        if task in self._in_use:
+            yield self._in_use[task]
+        else:
+            c = await self._read_connections.get()
+            try:
+                # record our connection in this dict to allow nested calls in
+                # the same task to use the same connection
+                self._in_use[task] = c
+                yield c
```

Comment:
Don't we still need a transaction of some form for multiple read statements to be taken from a single snapshot?

Reply:
I suppose it would be nice to offer this guarantee. I'm sure there are places that expects a consistent snapshot across multiple queries, but there may also be places that expect to see transactions immediately as they are committed.

On top of that, I'm not aware of a way to ask for a read-transaction in sqlite. As far as I can tell, a read-transaction is created (and closed) implicitly by a `SELECT` query.

I think, as well as disallowing nesting writes within reads, may require a more careful survey of all the places we use the database, before deciding what to do. (i.e. a follow-up PR)

Reply:
https://www.sqlite.org/lang_transaction.html
> If a write statement occurs while a read transaction is active, then the read transaction is upgraded to a write transaction if possible.

I believe you just start a transaction and it is a read transaction until otherwise upgraded.

Reply:
I thought you were asking about seeing a consistent snapshot of the database for the duration of the context manager. i.e. let the user acquire a connection and then read multiple times from the database and have all results be consistent. My understanding of a read-transaction in sqlite, is that a single SELECT query is a read transaction, when it ends, another SELECT query may see an updated state of the tables.

Reply:
That is what I was referring to, yes.  My understanding is that those are implicit read transactions that are created and closed when you haven't explicitly started a transaction prior to the `SELECT`.  Just the same as `INSERT` will create an implicit write transaction.

Yes, I was suggesting that the read connection should have savepoint handling just like the write connection so that when I acquire a connection and make two reads that they will be consistent with each other.  I both see your point about maybe some connections specifically don't want that and also I wonder if that's a questionable practice or at least something that is significantly less common.  I guess we may just have to survey the existing uses.  Maybe there would be one read context manager method for each mode.

As a detail, maybe the read connections wouldn't need the nesting behavior of savepoints, as opposed to just using a single transaction for the outermost request.  Or maybe it is simply enough once the savepoint is a context manager that we can just reuse that.

Reply:
I'm not sure creating a savepoint will necessarily make multiple queries operate on the same "snapshot" of the db, until the save point is closed. Does it work like that?